functorch/experimental/_cond.py

```diff
@@ -149,6 +158,100 @@ def cond_python_dispatcher(*args):
+        return True
+    except Exception as e:
+        raise e
+
+    input_nodes = set()
+    for node in gm.graph.nodes:
+        if node.op == "placeholder":
+            input_nodes.add(node)
+        if node.op == "call_function":
+            target = node.target
+            if isinstance(target, torch._ops.OpOverload) and target._schema.is_mutable:
+                for arg in node.args:
+                    if arg in input_nodes:
+                        return True
+
+    return False
+
+def _has_potential_branch_input_alias(branch, fake_inputs):
+    """
+    Dispatch-trace the branch with fake inputs and check if
+    producing graph has output aliasing the branch input. This is
+    bit restrictive as the branch must be traceable.
+    """
+    try:
+        gm = make_fx(branch)(*fake_inputs)
+    except UnsupportedMutationException:
+        # this can happen when nested cond is
+        # functionalized
+        return True
+    except Exception as e:
+        raise e
+
+    input_storages = set()
+    for node in gm.graph.nodes:
+        if node.op == "placeholder":
+            input_storages.add(StorageWeakRef(node.meta['val']._typed_storage()))
+
+    outs, _ = pytree.tree_flatten(gm(*fake_inputs))
+    for out in outs:
+        if isinstance(out, torch.Tensor) and StorageWeakRef(out._typed_storage()) in input_storages:
+            return True
+
+    return False
+
+
+
+@cond.py_impl(torch._C._functorch.TransformType.Functionalize)
+def cond_functionalize(interpreter, pred, true_fn, false_fn, inputs):
+    """
+    Functionalization implementation for torch.cond. Currently:
+      1. We don't allow any input mutation inside the branches
+      2. Our check for above condition is not exhaustive
+    """
+    reapply_views = interpreter.functionalize_add_back_views()
+    mode = 'mutations_and_views' if reapply_views else 'mutations'
+    # At this point, we will see functionalized tensors, so need to unwrap them first
+    unwrapped_inputs = _unwrap_all_tensors_from_functional(inputs, reapply_views=reapply_views)
+    unwrapped_pred = _unwrap_all_tensors_from_functional(pred, reapply_views=reapply_views)
+
+    functional_true_fn = functionalize(true_fn, remove=mode)
+    functional_false_fn = functionalize(false_fn, remove=mode)
+
+    with interpreter.lower():
+        fake_tensor_mode = FakeTensorMode()
+        with fake_tensor_mode as ft_mode:
+            for branch in [functional_true_fn, functional_false_fn]:
+                def convert(x):
+                    return ft_mode.fake_tensor_converter(ft_mode, x)
+                fake_inputs = pytree.tree_map_only(torch.Tensor, convert, unwrapped_inputs)
+                if _has_potential_branch_input_mutation(branch, fake_inputs):
+                    raise UnsupportedMutationException("One of torch.cond branch "
+                                                       "might be modifying the input!")
+            for branch in [true_fn, false_fn]:
```

Comment:
nit: isn't this for loop duplicated above on line 237?

Reply:
Ohh the above one is running over functionalized branch, this one is for non-functional branch. This is because when we functionalize, the storage info gets lost because we use view_copy instead of view.